Added a BDD test to prove the bug, with code to fix it.

fjank opened the pull request.

fjank committed:
Added a BDD test to prove the bug.

netbrain reviewed (dismissed):
Looks like a valid test.

fjank committed:
Added nasty code to fix the new test, and hopefully any other tests that is added.

netbrain commented on src/main/java/github/workflow/kata/Grid.java:
```diff
@@ -6,10 +6,69 @@
+
+    public void nextGeneration() {
+        // so, lets try to implement the actual code.
+        // not so happy with the implementation, it's "a bit" dirty, but it works. Refactor this later.
```
later is never

fjank replied:
on it

netbrain commented on src/main/java/github/workflow/kata/Grid.java:
```diff
@@ -6,10 +6,69 @@
  */
 public class Grid {
-    private final String gridDef;
+    private String gridDef;
```
There are probably better data structures to represent a grid.

netbrain replied:
would recommend using either a single array with a stride parameter or a 2xdimensional array

fjank replied:
going for 2xdim array, strides are a bit cumbersome.

netbrain commented on src/main/java/github/workflow/kata/Grid.java:
```diff
@@ -6,10 +6,69 @@
+            for (int col = 0; col < cols.length; col++) {
+                char ch = cols[col];
+                // Rule: As a live cell I will die if I have fewer than two live neighbours
+                if (ch == '*') {
+                    // check the surrounding 9 cells.
+                    int liveCount = 0;
+                    int startRow = Math.max(0, row - 1);
+                    int endRow = Math.min(rows.length-1, row + 1);
+                    int startCol = Math.max(0, col - 1);
+                    int endCol = Math.min(cols.length-1, col + 1);
+                    for (int i = startRow; i <= endRow; i++) {
+                        for (int j = startCol; j <= endCol; j++) {
+                            // skip ourselves.
+                            if (i == row && j == col) {
+                                continue;
+                            }
+                            if (rows[i].toCharArray()[j] == '*') {
+                                liveCount++;
+                            }
+                        }
+                    }
+                    if (liveCount < 2) {
+                        // cell will die.
+                        newDef += '.';
+                    } else {
+                        // cell will still be alive.
+                        newDef += '*';
+                    }
+                } else {
+                    // cell is dead, it will also be dead next round.
+                    newDef += '.';
+                }
```
This can probably be extracted to one or more functions

fjank replied:
oh yes, that seems appropriate.

netbrain commented on src/main/java/github/workflow/kata/Grid.java:
```diff
@@ -6,10 +6,69 @@
+        // so, lets try to implement the actual code.
+        // not so happy with the implementation, it's "a bit" dirty, but it works. Refactor this later.
+        String newDef = "";
+        String[] rows = gridDef.split("\n");
+        for (int row = 0; row < rows.length; row++) {
+            char[] cols = rows[row].toCharArray();
+            for (int col = 0; col < cols.length; col++) {
+                char ch = cols[col];
```
```suggestion
        for (String row : gridDef.split("\n")) {
            for (char ch :  row.toCharArray()) {
```

fjank replied:
it looks differend after the refactorings I have done.

fjank committed:
Refactored the grid class content

now use a char[][] as grid, better than before, but still a bit "raw".
Can probably be more OO, but that can be fixed at a later point in time,
when more rules are implemented.
